functions/access.js

```diff
@@ -0,0 +1,142 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require("./auth");
+const { uuid } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  console.log("asdasd");
+  if (!uid) return false;
+
+  console.log("asdasd");
+  const venueRef = admin.firestore().collection("venues").doc(venueId);
+  const accessRef = admin.firestore().collection("accessgranted").doc(uid);
+
+  try {
+    console.log("asdasd");
+    await admin.firestore().runTransaction(async (t) => {
+      const venue = await t.get(venueRef);
+      const granted = await t.get(accessRef);
+
+      if (!venue.exists) {
+        throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+      }
+      if (!granted.exists || !granted.tokens) {
+        return false;
+      }
+
+      console.log("v", venue);
+      console.log("g", granted);
+      if (Object.keys(granted.tokens).includes(token)) {
+        // Record that the token was checked
+        if (!granted.tokens[token].usedAt) {
+          granted.tokens[token].usedAt = [];
+        }
+        granted.tokens[token].usedAt.push(Date.now());
+        t.update(accessRef, granted);
+
+        return true;
+      } else {
+        return false;
+      }
+    });
+  } catch (e) {
+    console.log("Transaction failure:", e);
+    return false;
+  }
+
+  return false;
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = await getAccessDoc(venueId, "password");
+  if (!access.exists || !access.data().password) {
+    return false;
+  }
+  return passwordsMatch(access.data().password, password);
+};
+
+const isValidEmail = async (venueId, email) => {
+  const access = await getAccessDoc(venueId, "emails");
+  if (!access.exists || !access.data().emails) {
+    return false;
+  }
+  return access.data().emails.includes(email.trim().toLowerCase());
+};
+
+const isValidCode = async (venueId, code) => {
+  const access = getAccessDoc(venueId, "code");
+  if (!access.exists || !access.data().codes) {
+    return false;
+  }
+  return access.data().codes.includes(code.trim());
```

Comment:
> Will the stored codes be guaranteed to be .trim()?

I can't say for sure. I believe the values are currently entered manually in the firebase.

Reply:
Then we should do the map and/or find thing so that we can normalise both sides of it here, and not need to be concerned that the data we are getting is already ‘perfectly formatted’

Reply:
@denisdimitrov This looks like it still needs to be done?

Reply:
@mike-lvov @denisdimitrov As above, I believe not. So this is tech-debt now that this has been merged. Also called out in https://github.com/sparkletown/internal-sparkle-issues/issues/312 (but a specific backlog issue/fix should be done for this, as that issue covers a huge amount of other things, of which this is just one aside mentioned)

Reply:
Backlogged in https://github.com/sparkletown/internal-sparkle-issues/issues/313